core/node/vm_runner/src/output_handler.rs

```diff
@@ -0,0 +1,541 @@
+            .field("pool", &self.pool)
+            .field("io", &self.io)
+            .field("factory", &self.factory)
+            .finish()
+    }
+}
+
+impl<Io: VmRunnerIo + Clone, F: OutputHandlerFactory> ConcurrentOutputHandlerFactory<Io, F> {
+    /// Creates a new concurrent delegator factory using provided Postgres pool, VM runner IO
+    /// and underlying output handler factory.
+    ///
+    /// Returns a [`ConcurrentOutputHandlerFactoryTask`] which is supposed to be run by the caller.
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        io: Io,
+        factory: F,
+    ) -> (Self, ConcurrentOutputHandlerFactoryTask<Io>) {
+        let state = Arc::new(DashMap::new());
+        let task = ConcurrentOutputHandlerFactoryTask {
+            pool: pool.clone(),
+            io: io.clone(),
+            state: state.clone(),
+        };
+        (
+            Self {
+                pool,
+                state,
+                io,
+                factory,
+            },
+            task,
+        )
+    }
+}
+
+#[async_trait]
+impl<Io: VmRunnerIo, F: OutputHandlerFactory> OutputHandlerFactory
+    for ConcurrentOutputHandlerFactory<Io, F>
+{
+    async fn create_handler(
+        &mut self,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<Box<dyn StateKeeperOutputHandler>> {
+        let mut conn = self.pool.connection_tagged(Io::name()).await?;
+        let latest_processed_batch = self.io.latest_processed_batch(&mut conn).await?;
+        let last_processable_batch = self.io.last_ready_to_be_loaded_batch(&mut conn).await?;
+        drop(conn);
+        anyhow::ensure!(
+            l1_batch_number > latest_processed_batch,
+            "Cannot handle an already processed batch #{} (latest is #{})",
+            l1_batch_number,
+            latest_processed_batch
+        );
+        anyhow::ensure!(
+            l1_batch_number <= last_processable_batch,
+            "Cannot handle batch #{} as it is too far away from latest batch #{} (last processable batch is #{})",
+            l1_batch_number,
+            latest_processed_batch,
+            last_processable_batch
+        );
+
+        let handler = self.factory.create_handler(l1_batch_number).await?;
+        let (sender, receiver) = oneshot::channel();
+        self.state.insert(l1_batch_number, receiver);
+        Ok(Box::new(AsyncOutputHandler {
+            internal: Some(OutputHandlerState::Running { handler, sender }),
+        }))
+    }
+}
+
+enum OutputHandlerState {
+    Running {
+        handler: Box<dyn StateKeeperOutputHandler>,
+        sender: oneshot::Sender<BoxFuture<'static, anyhow::Result<()>>>,
+    },
+    Finished,
+}
+
+impl Debug for OutputHandlerState {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("OutputHandlerState").finish()
+    }
+}
+
+#[derive(Debug)]
+struct AsyncOutputHandler {
+    internal: Option<OutputHandlerState>,
```

Comment:
Using an `Option` looks redundant here. You can use `mem::replace(&mut self.internal, OutputHandlerState::Finished)` in `handle_l1_batch()` instead. This would allow to unite `OutputHandlerState` and `AsyncOutputHandler` in a single type.